### PR TITLE
Failing test case for the blocking problem

### PR DESCRIPTION
**Pull request by NorfairKing**
@robinp This PR is to address the blocking problem.
I hope it's solvable but I'll add some test cases first.
In particular, I'd like to keep the fairness property if possible.

### CODE REVIEW

**Comment by NorfairKing** [comment]
@robinp I think I have a solution to the blocking problem. (with tests! :D)
What do you think about it?

File: token-limiter-concurrent/test/Control/Concurrent/TokenLimiter/ConcurrentSpec.hs [hunk 1]
```diff
@@ -189,6 +236,26 @@ spec = do
...
+                  { tokenLimitConfigInitialTokens = 0,
+                    tokenLimitConfigMaxTokens = 1,
+                    tokenLimitConfigTokensPerSecond = 50
+                  }
+          limiter <- makeTokenLimiter config
+          let l :: [Int]
+              l = [1 .. numberOfThreads]
+          -- The threads start waiting in order, so they need to wake up in order
+          forConcurrently_ l $ \ix -> do
+            threadDelay (10_000 * ix)
```
Comment (robinp) [inline token-limiter-concurrent/test/Control/Concurrent/TokenLimiter/ConcurrentSpec.hs:253]: Didn't check these tests, random note: I found it can be useful to execute `stress` utility with CPU/memory/IO hogs while performing the concurrent tests repeatedly, scheduling errors are much proner to appear under high load.

**Review by robinp — COMMENTED** [review]
Some review notes based on gut feeling, take with grain of salt.

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs [hunk 2]
```diff
@@ -81,27 +83,39 @@ makeTokenLimiter tokenLimiterConfig = do
...
 canDebit :: TokenLimiter -> Word64 -> IO Bool
-canDebit TokenLimiter {..} debit = withMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
-  now <- getMonotonicTimeNSec
-  let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
-  let enoughAvailable = currentCount >= debit
-  pure enoughAvailable
+canDebit TokenLimiter {..} debit = do
+  -- If any thread is currently waiting, we definitely can't debit any tokens
+  -- currently.
+  currentlyWaiting <- isEmptyMVar tokenLimiterWaitLock
```
[the remark this answers is not in the document]
Comment (NorfairKing) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:89]: > The problem is, the two MVar's are independent, so even if isEmptyMVar is False here, there isn't any guarantee that by the next line (withMVar) the tokenLimiterLastServiced MVar wouldn't already be taken.

This is a fair point.
In that case the output would still be correct but the call would still (sadly) block.

> If we want to stay with an MVar-only solution

Not necessarily. How would you suggest to do this?
I think you can't `threadDelay` in an `stm`...

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs [hunk 3]
```diff
@@ -133,13 +147,13 @@ waitDebitObserve TokenLimiter {..} debit = modifyMVar tokenLimiterLastServiced $
...
-              * fromIntegral extraTokensNeeded
-              -- fromIntegral :: Word64 -> Double
-              / fromIntegral (tokenLimitConfigTokensPerSecond tokenLimiterConfig)
+              * (fromIntegral :: Word64 -> Double) extraTokensNeeded
+              / (fromIntegral :: Word64 -> Double) (tokenLimitConfigTokensPerSecond tokenLimiterConfig)
+
       let microsecondsToWait = ceiling microsecondsToWaitDouble
-      -- threadDelay guarantees that _at least_ the given number of microseconds will have passed.
-      threadDelay microsecondsToWait
+      withMVar tokenLimiterWaitLock $ \() ->
```
[the remark this answers is not in the document]
Comment (NorfairKing) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: @robinp That significantly changes the semantics.
We want the `WaitLock` to only be taken when actual waiting is occurring, otherwise the results of the other functions (that return false if this lock is taken) are different.

[the remark this answers is not in the document]
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: Well, once the `waitDebit` function is called, there _is_ some actual wait that would be occurring (given a concurrent caller), at most at the limit it is zero-ish (if the current token count suffices). So I don't think it is meaningful to distinguish the zeroish-wait case from the nonzeroish-wait case, from the perspective of concurrent debitors.

So as long as tryTakeMVar is used for canDebit / tryDebit (and the semantic change of "False" as described in other comment is accepted), it should be ok to take the lock here at the beginning.

But I think we are designing a bit in a vacuum - it is hard to tell what is reasonable behavior from the canDebit / tryDebit functions without knowing how the callers would use the result. For example, two main scenarios I can imagine (but I'm not very imaginative now):

A) Caller would anyway need to perform the action after the rate limit. Then not much use in canDebit/tryDebit, just grit teeth and do a waitDebit.

B) Caller would shed actions that can't be immediately actioned.

Maybe the return values of canDebit/tryDebit shouldn't be Bool, but a tri-state of `(Possible)Success|NotEnoughTokens|OtherActionIsWaiting` (where the currently proposed change's `False` just conflates the two latter), and let the caller decide up to its taste how to proceed.

(Btw, something like a `tryModifyMVar` helper that could be written could be utilized to provide this separation)

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:89]: You can do using `unsafeIoToSTM`, and `threadDelay` is benign here, but a problem is that after the current waiter finishes, a random other blocked STM waiter would be restarted, without fairness guarantee. So starving a waiter is possible. I don't immediately see a way to solve that within STM (but it doesn't mean it can't).

[the remark this answers is not in the document]
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:89]: Hm, maybe separating the operation to two stages could work: first stage would be an STM action that would add a unique id (myThreadId?) to a STM queue. This would be performed with atomically. Second stage would then be an STM action that reads from the queue, and retries unless the read value is its own unique. This way there's an established FIFO order on the actions. The second stage could also do the accounting of the tokens, as these are all STM actions, they compose nicely.

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (NorfairKing) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: @robinp This has been in my head all evening now.
You've just given me the most useful insight.

We don't need `canDebit` at all, I think it was a misfeature.

I'll tell you the exact behaviour I want.
I have a website (https://social-dance.today) that needs to do geocoding.
There are two geocoding providers:

1. OpenStreetMaps' Nominatim, which is free, but hard-rate-limited to 1 request per second.
2. Google Maps, which is paying.

I have two use cases: Geocoding ASAP (for web requests) and Geocoding whenever (for batch processes).

In the ASAP version, I want to use `tryDebit` to use OpenStreetMaps if the rate limit allows that, and google maps otherwise, to save money.
In the "whenever" version, I want to use `waitDebit` to always OpenStreetMaps and save time that way.

[the remark this answers is not in the document]
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: If we define "could have debited" as False if the MVar is currently taken, then the `tryModifyMVar` route for `tryDebit` is sufficient.

If that doesn't fly, then sounds like one needs to introduce a "max tolerated timeout" for `tryDebit` to make its decision (and then either enforce it by some combination of takeMVar + Async.race), but it sounds like premature optimization to avoid the small race window of MVar contention in the enough-tokens case.

Probably the API cost you would have to pay over lifetime for those contention-originated calls is less than the code review time cost :)

[the remark this answers is not in the document]
Comment (NorfairKing) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: @robinp Where do you see `tryModifyMVar` I can't seem to find it?

[the remark this answers is not in the document]
Comment (NorfairKing) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:154]: Aha, was able to write it based on another codebase:

```                                       
tryModifyMVar :: MVar a -> (a -> IO (a, b)) -> IO (Maybe b)               
tryModifyMVar m io =                               
  uninterruptibleMask $ \restore -> do   
    maybea <- tryTakeMVar m                                                  
    forM maybea $ \a -> do                               
      (a', b) <-                                                    
        restore (io a)                                                                                    
          `onException` putMVar m a                                        
      putMVar m a'                       
      return b
```

**Comment by NorfairKing** [comment]
Ok @robinp Please have a look at the latest commit. It should do what you had in mind using `tryModifyMVar`.

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs [hunk 4]
```diff
@@ -73,54 +76,41 @@ makeTokenLimiter tokenLimiterConfig = do
...
-    then do
-      let newCount = currentCount - debit
-      pure ((now, newCount), True)
-    else pure ((lastServiced, countThen), False)
+tryDebit TokenLimiter {..} debit =
+  fmap (fromMaybe False) $
+    tryModifyMVar tokenLimiterLastServiced $ \(lastServiced, countThen) -> do
+      now <- getMonotonicTimeNSec
+      let currentCount = computeCurrentCount tokenLimiterConfig lastServiced countThen now
+      let enoughAvailable = currentCount >= debit
```
Comment (robinp) [inline token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:93]: optional nit: for multiple `let`s, can drop the non-first let (while keeping the indent).

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs [hunk 5]
```diff
@@ -73,54 +76,41 @@ makeTokenLimiter tokenLimiterConfig = do
...
+      if enoughAvailable
+        then do
+          let newCount = currentCount - debit
+          pure ((now, newCount), True)
+        else pure ((lastServiced, countThen), False)
```
Comment (robinp) [inline token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:98]: optional nit: as both branches end in pure, can pull up the pure in front of the if

**Review by robinp — COMMENTED** [review]
Looks generally good!

Mostly nits, apply up to personal taste.

File: token-limiter-concurrent/test/Control/Concurrent/TokenLimiter/ConcurrentSpec.hs (hunk 1 above)
Comment (NorfairKing) [reply token-limiter-concurrent/test/Control/Concurrent/TokenLimiter/ConcurrentSpec.hs:195]: Great tip!

I tried this for a while and didn't see anything fail, but I probably didn't stress the system in exactly the right way to cause issues.

**Comment by NorfairKing** [comment]
@robinp Thanks again for the great code review, I really appreciate it!

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs (hunk 4 above)
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs:93]: Ah, nice trick. Thanks!

File: token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs [hunk 6]
```diff
@@ -176,3 +163,14 @@ computeCurrentCount TokenLimitConfig {..} lastServiced countThen now =
    in if willOverflow
         then tokenLimitConfigMaxTokens
         else min tokenLimitConfigMaxTokens totalCount
+
+tryModifyMVar :: MVar a -> (a -> IO (a, b)) -> IO (Maybe b)
+tryModifyMVar m io =
+  uninterruptibleMask $ \restore -> do
```
[the remark this answers is not in the document]
Comment (robinp) [reply token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs]: I think https://www.fpcomplete.com/haskell/tutorial/exceptions/ is a great read about this still. If you would like to deal with Haskell concurrency for some long time, I suggest strongly investing in understanding this (and the above linked Control.Exception doc on interruptible).

TLDR in Haskell a thread running some code can be excepted at "any" point (well, if it doesn't malloc, and runs in a tight computation loop, then it can't unless you explicitly use `yield`... but that's rare). 

(What is an interrupt? More specifically it is an async Exception, which is just an exception thrown asynchronously, that is to an other thread (like with throwTo). The Async lib for example uses this under the hood of `cancel` and all its combinators like concurrently etc)

Now, if you have piece of code that needs to uphold some invariant (like with modifyMVar, I either leave the MVar as-is, or write back the modified content, but never accidentally leave the MVar empty, which would then likely deadlock subsequent users), then you need to think if/where interrupts can occur. Usually what you want, you mask interrupts (that is, disable receiving them) for the setup of your handling code, and restore interruptibility for the action (once the handlers are set up).

Now, 'mask' disables interrupts, _but_ they can still happen at specific points (the "interruptible" points IIRC), which usually happen around "blocking" operations (from the app perspective - the runtime itself doesn't block but yields control to other stuff). This is to prevent deadlocks and whatever.

'uninterruptibleMask' really masks all interrupts, but then it means if your masked computation gets in a loop or something, there's no way to kill it. For example, in server code you usually want to serve requests from users in an Async and put some hard upper bound on execution time (via race + threadDelay). But if the computation is uninterruptibly masked, then it can't receive the async exception that kills it, and you are deadlocked. So don't use that.

(unliftio, rio use uninterruptible mask in their versions of `finally` etc, to guarantee completion of resource cleanup, taking the side of rather deadlocking than leaking resources. I lean to this approach as well, but it has some tricky downsides in some situations... it is kind of all-in, don't mix different semantics, either stick to base or full-in on unliftio)
